### PR TITLE
CMake build fixes

### PR DESCRIPTION
**Pull request by AlexMax**
This pull request fixes the problematic packaging of DLL files under MinGW.  Please let me know if you have any questions.

### CODE REVIEW

**Comment by chungy** [comment]
On testing it, it bombs out just on the compilation phase, trying to copy a file that doesn't exist:

```
Error copying file (if different) from "/usr/i686-w64-mingw32/bin/libopus-0.dll" to "/home/chungy/Games/doom/woof/build/Source".
make[2]: *** [Source/CMakeFiles/woof.dir/build.make:1177: Source/woof.exe] Error 1
make[2]: *** Deleting file 'Source/woof.exe'
make[2]: Leaving directory '/home/chungy/Games/doom/woof/build'
make[1]: *** [CMakeFiles/Makefile2:207: Source/CMakeFiles/woof.dir/all] Error 2
make[1]: Leaving directory '/home/chungy/Games/doom/woof/build'
make: *** [Makefile:152: all] Error 2
```

**Comment by AlexMax** [comment]
We were discussing this in #61 about how using mingw's libraries as they are supplied by pacman is unworkable going forward.  However, I think that one ought to be able to compile the program using mingw libraries, just not package them.  I will amend the build files and update this PR.

**Comment by AlexMax** [comment]
@chungy , are your mingw SDL libraries out of date?  I was able to properly build the program with my MSYS2 installation, even if the libraries being moved to the build directory was kind of a useless gesture.

**Comment by chungy** [comment]
With due respect, all this just seems like a very fragile solution that `cp-with-libs` solved succinctly, and is used with good success in both the <a href="https://github.com/chocolate-doom/chocolate-doom/tree/master/pkg/win32">chocolate-doom</a> and <a href="https://github.com/Doom-Utils/deutex/tree/master/pkg.w32">deutex</a> projects (yeah they are kind of wrapped around autotools, but the program itself doesn't actually have dependence on autotools).

> @chungy , are your mingw SDL libraries out of date? I was able to properly build the program with my MSYS2 installation, even if the libraries being moved to the build directory was kind of a useless gesture.

No, I don't think they are:
```
$ pacman -Q mingw-w64-sdl2 mingw-w64-sdl2_net mingw-w64-sdl2_mixer 
mingw-w64-sdl2 2.0.10-3
mingw-w64-sdl2_net 2.0.1-1
mingw-w64-sdl2_mixer 2.0.4-2
```

**Comment by AlexMax** [comment]
`cp-with-libs` does have a dependency on `objdump`, which you are not assured of having unless you have a GNU compiling environment installed.  The other problem is that the strategy of blanketly copying files out of the mingw directory leads to madness, as I found out first-hand [here](https://github.com/fabiangreffrath/woof/issues/61#issuecomment-589894786).

If you know why that's happening and how to prevent dumping a truckload of dll's into the distribution, I'm all ears.  But in lieu of that, we decided to simply assume that if you intend to package the software, you'll use the official development libraries downloaded from libsdl.org.

**Comment by chungy** [comment]
> The other problem is that the strategy of blanketly copying files out of the mingw directory leads to madness [..] If you know why that's happening and how to prevent dumping a truckload of dll's into the distribution, I'm all ears.

`cp-with-libs` does exactly that, though! If I checkout b66f8b89f7b3a15ce03c181d15fbec4dc00e6f20 and build a Windows exe (32-bit version; 64-bit doesn't require libgcc*.dll for whatever reason), this is all I get:
```
$ unzip -l woof-0.9.0-win32.zip \*.dll
Archive:  woof-0.9.0-win32.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
    93696  2020-02-25 17:32   libgcc_s_sjlj-1.dll
    70656  2020-02-25 17:32   libwinpthread-1.dll
  1021440  2020-02-25 17:32   SDL2.dll
   116736  2020-02-25 17:32   SDL2_mixer.dll
    20992  2020-02-25 17:32   SDL2_net.dll
---------                     -------
  1323520                     5 files
```

**Comment by AlexMax** [comment]
I'm telling you my experience of what I saw when I copied the files out of the directory by hand.  I would try to start the program (from outside MSYS2), I would see which DLL's that Windows would complain about - presumably errors that prevented the program from starting - and then copy those specific DLL's into the directory manually.  I stopped when I started seeing oddball DLL files that I didn't recognize.  Do you have an explanation for why this happened, and how `cp-with-libs` avoids this problem?

**Comment by chungy** [comment]
> Do you have an explanation for why this happened

Not in particular. Might be related to however your packages were built.

> and how `cp-with-libs` avoids this problem?

It simply calls on `objdump` to read the dependent libraries of every PE file, copying the required ones into a directory, excluding a few that basically every Windows system have in `C:\Windows\System32`. I'll copy an example of a build I just did, it seems to run fine for me without having to copy any other random files, but that is only in Wine; I don't have a Windows system to test it on the native OS.
https://chungy.keybase.pub/woof-0.9.0-win32.zip

**Comment by AlexMax** [comment]
> Not in particular. Might be related to however your packages were built.

The problematic packages I am using came from MSYS2's MinGW installation, installed using `pacman`.  Perhaps the libraries you have access to were more self-contained than those.

> It simply calls on `objdump` to read the dependent libraries of every PE file, copying the required ones into a directory, excluding a few that basically every Windows system have in `C:\Windows\System32`.

I am aware of how the program works by glancing at its source, but I didn't know if there was some additional magic on top of that, so it sounds like the answer to my question is that "it doesn't avoid the problem".

That said, having to guess the names of the DLL files is indeed a fragile ffix, but on the other hand I do not want to introduce a hard dependency on a UNIX tool just to copy DLL files to a destination directory on Windows.  What I think I will do is use `cp-with-libs` in instances where `objdump` is detected, and revert to guessing the filenames if it's not, unless we can come up with a clever workaround to avoid `objdump`.  This work will not be finished tonight.

**Comment by fabiangreffrath** [comment]
> What I think I will do is use `cp-with-libs` in instances where `objdump` is detected, and revert to guessing the filenames if it's not, unless we can come up with a clever workaround to avoid `objdump`. 

This sounds like the most reasonable solution! When I agreed that it would be best to use the officially released SDL libraries for packaging, I didn't have in mind that this would be left as the only solution. Indeed, if we rely on `cp-with-libs` if its requirements are found, we'll have a solution that works with all SDL libraries, regardless if they are the official ones or the ones packaged in MSYS2. And we can still have a reasonable fallback on systems without `objdump`.

> This work will not be finished tonight.

No problem, take your time. We are not in a hurry.

Thank you for all the effort you spent on this so far!

**Comment by AlexMax** [comment]
Okay, so `cp-with-libs` might have an issue too.

I tried to use `cp-with-libs` on the SDL development libraries downloaded from SDL's website, and it only copied the main `SDL_mixer.dll` file, along with `SDL.dll`.  None of the other files were copied.

Doing a little digging, I realized that SDL_mixer allows loading these external DLL files dynamically using `SDL_LoadObject`, which means that they would not be listed in `objdump -p` at all, which is what `cp-with-libs` uses to figure out what files are necessary to copy.  We could also try parsing out the contents of `strings` which does contain the exact name of every referenced dll file, but that seems fragile.  I suppose when using MSYS and Linux distro packages the libraries are directly linked against, but not for downloaded development libraries from SDL's website.

Crud.  This isn't even the build system messing with us, it's SDL being cute about loading libraries.

- - -

At this point, I almost feel like we should be opinionated about which versions of libraries we compile with Woof on Windows to avoid the headaches surrounding figuring out which DLL files to package.  In a past (non-Doom) project, I have used CMake's [ExternalProject](https://cmake.org/cmake/help/latest/module/ExternalProject.html) support to download, compile, and automatically link against libraries.  It worked shockingly well.  We could possibly use this functionality to automatically download and link against the proper development libraries for Windows builds.

This would actually have a significant advantage where no user intervention would be required to properly build on Windows - no downloading libraries separately or anything.  What do you think, @fabiangreffrath ?

**Comment by fabiangreffrath** [comment]
> Doing a little digging, I realized that SDL_mixer allows loading these external DLL files dynamically 

So, strictly speaking `cp-with-libs` is doing the correct thing and does only copy the files which are actually necessary to use the library - i.e. the ones the library was linked against. All other files are optional and should be treated as such.

> At this point, I almost feel like we should be opinionated about which versions of libraries we compile with Woof on Windows to avoid the headaches surrounding figuring out which DLL files to package. In a past (non-Doom) project, I have used CMake's [ExternalProject](https://cmake.org/cmake/help/latest/module/ExternalProject.html) support to download, compile, and automatically link against libraries. 

Does this mean we would have to compile SDL2_Mixer with all its dependencies down to libogg, libvorbis, libopus, libflac, libmpg123 and all this stuff ourselves? 

What about fluidsynth? I'd like to have this as a optional `dlopen()`ed dependency of SDL2_Mixer. I don't want to carry it around with all its own dependencies like libreadline and libglib, but if it is somehow provided at run-time, I'd like to be able to use it for soundfont rendering.

> This would actually have a significant advantage where no user intervention would be required to properly build on Windows - no downloading libraries separately or anything. What do you think, @fabiangreffrath ?

I think I would prefer it as simple as this: If all pre-requisites for `cp-with-libs` are available, use this one. If not, copy the SDL libraries and all the libraries found in the same directory.

**Comment by AlexMax** [comment]
> > Doing a little digging, I realized that SDL_mixer allows loading these external DLL files dynamically
> 
> So, strictly speaking `cp-with-libs` is doing the correct thing and does only copy the files which are actually necessary to use the library - i.e. the ones the library was linked against. All other files are optional and should be treated as such.

It might be acceptable if all we care about is MIDI support, but as soon as you consider supporting other music that requires one of those third-party libraries we will need to have this conversation all over again.

> > At this point, I almost feel like we should be opinionated about which versions of libraries we compile with Woof on Windows to avoid the headaches surrounding figuring out which DLL files to package. In a past (non-Doom) project, I have used CMake's [ExternalProject](https://cmake.org/cmake/help/latest/module/ExternalProject.html) support to download, compile, and automatically link against libraries.
> 
> Does this mean we would have to compile SDL2_Mixer with all its dependencies down to libogg, libvorbis, libopus, libflac, libmpg123 and all this stuff ourselves?

No, we can use the development libraries that already come prepackaged with the proper libraries to link against.  All it would do is automate downloading and extraction to a known location in the build directory that we could then find with our finder script, the "build" step can be skipped, or if it can't the same functionality can be done with a custom target trivially.

> What about fluidsynth? I'd like to have this as a optional `dlopen()`ed dependency of SDL2_Mixer. I don't want to carry it around with all its own dependencies like libreadline and libglib, but if it is somehow provided at run-time, I'd like to be able to use it for soundfont rendering.

That's a good question - I don't know if SDL_mixer as provided by SDL's site allows linking against fluidsynth dynamically.

> I think I would prefer it as simple as this: If all pre-requisites for `cp-with-libs` are available, use this one. If not, copy the SDL libraries and all the libraries found in the same directory.

Okay, I'll do that as a fallback.

**Comment by fabiangreffrath** [comment]
> No, we can use the development libraries that already come prepackaged with the proper libraries to link against. All it would do is automate downloading and extraction to a known location in the build directory that we could then find with our finder script, the "build" step can be skipped, or if it can't the same functionality can be done with a custom target trivially.

Alright, this seems to make a lot of sense.

> That's a good question - I don't know if SDL_mixer as provided by SDL's site allows linking against fluidsynth dynamically.

Me neither. The local MSYS2 package for SDL2_Mixer that I currently use does, though, but this is a solution that I'd like to get rid as at the earliest convenience. So, this is clearly second priority, let's get a clean solution for Windows targers first and then clean out the detail.

> Okay, if we can't come to an agreement on the automatic download (the question that I do not consider resolved quite yet), I'm happy to do this as a fallback.

I am fine with automatic downloading if it turns out as easy as you outlined above.

**Comment by fabiangreffrath** [comment]
Ah, and BTW, what about ZLib?

**Comment by fabiangreffrath** [comment]
Oh, I just see that ZLib is part of the SDL2_Image distribution, so including this would solve two problems at once. Also, this is with the officially released SDL-libraries:
```
$ objdump -x SDL2_image.dll | grep DLL
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll
        DLL Name: SDL2.dll

$ objdump -x SDL2_mixer.dll | grep DLL
        DLL
 vma:            Hint    Time      Forward  DLL       First
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll
        DLL Name: SDL2.dll
        DLL Name: WINMM.DLL
```

So, they are not linked against any of the other libraries that they bundle, which means we could easily strip out e.g. libjpeg, libtiff and libwebp from the Woof distribution without breaking the SDL libraries's dependency chains.

**Comment by AlexMax** [comment]
> Oh, I just see that ZLib is part of the SDL2_Image distribution, so including this would solve two problems at once.

Not necessarily.  SDL2_image only contains the .dll file for zlib, and does not come with zlib headers that are necessary to actually write code against zlib itself.

I see that you added this:

```cmake
check_library_exists(z inflate "" z_FOUND)
```

In general, `check_library_exists` is only really appropriate for libraries that are supposed to be supplied by the compiler, things that you either use the version that comes with the system or nothing at all.  This is why I use it for `dl` and `m`.  For external libraries like zlib, where you might reasonably supply your own version or reasonably expect to have to locate it on Windows, you need to write a finder package, like I did with SDL.

Luckily, some libraries are so common that CMake comes with finders built-in.  Zlib happens to be one of those.  You can find the documentation on FindZLIB [here](https://cmake.org/cmake/help/latest/module/FindZLIB.html).

However, we now have a serious problem - we have a dependency that's needed both by itself and as a transitive dependency of SDL_image.  On Linux, this is fine, because most of the time everything is in `/usr/lib` and it's all connected together anyway.  But on Windows, the path is less clear, especially since we're trying to use development libraries to avoid problems with MSYS2, and those development libraries come with a mystery zlib1.dll.

Frankly, this is why I wanted to use stb_image_writer, to avoid problems with transitive dependencies like these.  I will need a little more thought to figure out the best approach.

**Comment by fabiangreffrath** [comment]
Sigh. Wasn't CMake supposed to make things easier?

**Comment by fabiangreffrath** [comment]
> Frankly, this is why I wanted to use stb_image_writer, to avoid problems with transitive dependencies like these. I will need a little more thought to figure out the best approach. Not an insurmountable problem by any means, just that more thought is needed.

Well, I am not going to base my choice of image library on CMake's ability to cope with the ZLib dependency.

**Comment by fabiangreffrath** [comment]
If we assume that ZLibs are always absolutely ABI-compatible, which I think we can, would that help? As in, it doesn't matter which one is installed, the system one or the one from SDL2_Image, even if one overwrites the other upon installation.

**Comment by AlexMax** [comment]
> Sigh. Wasn't CMake supposed to make things easier?

This is not a CMake issue.  This is a C library issue.

`automake` and `cp-with-libs` have thus far limited you to mingw+MSYS2 on Windows and mingw cross-compiling on Linux.  Now that CMake allows you to use Visual Studio and mingw on Windows without MSYS2, things are more complicated.  These problems would have to be solved with CMake, Meson, Premake, or whatever else was chosen.

**Comment by AlexMax** [comment]
Okay, so just letting you know where the branch is, SDL_image is now seeked and linked against just like the other three SDL libraries, and zlib is currently sought out on Linux and automatically compiled on Windows, both using MSVC and with a mingw cross-build.

Packaging works too.  SDL_image is included, along with libpng and its copy of zlib1.dll.  Unzipping the archive and running the software results in no crashes.

I have not tested this setup in a native MSYS2 environment yet - I reformatted recently, so I don't have it on my heard drive yet.  Will do some testing tomorrow.

**Comment by fabiangreffrath** [comment]
@AlexMax This sounds all great, what a vast improvement, thank you very much!

Please let me know when you have checked the build system in MSYS2 so I can merge this branch ASAP.

File: Source/p_extnodes.c [hunk 1]
```diff
@@ -308,9 +302,7 @@ void P_LoadNodes_ZDBSP (int lump, boolean compressed)
 	// release the original data lump
 	W_CacheLumpNum(lump, PU_CACHE);
 	free(zstream);
-#else
 	I_Error("P_LoadNodes: Compressed ZDBSP nodes are not supported!");
```
Comment (fabiangreffrath) [inline Source/p_extnodes.c]: This line need to be removed if we unconditionally build with ZLIb support.

Comment (AlexMax) [reply Source/p_extnodes.c]: Got it.

**Comment by AlexMax** [comment]
I have been unable to test with MSYS2, as the installer on their home page is not working since yesterday.  I have tested it with Visual Studio and MinGW on Linux.

**Comment by chungy** [comment]
> MinGW on Linux.

Not working for me:
```
CMake Error at Source/CMakeLists.txt:103 (add_executable):
  Target "woof" links to target "SDL2::image" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
```

(I'm also personally feeling deeply uncomfortable with this automatic downloading business... :/)

**Comment by AlexMax** [comment]
I finally managed to install MSYS2 and it does appear to download and compile everything correctly.  Perhaps try clearing your build directory?

> (I'm also personally feeling deeply uncomfortable with this automatic downloading business... :/)

All of the library archives have their SHA256 hash embedded in the CMakeLists, to ensure that the thing you are downloading is what is intended.  You did remind me that I had to do the same thing with zlib though, so I went ahead and did that too.

**Comment by chungy** [comment]
> Perhaps try clearing your build directory?

Didn't have one in the first place. Just get the error.

**Comment by fabiangreffrath** [comment]
I have just tested this in my MSYS2 environment and it worked as expected! The only nit-pick that I have is that we should probably get rid of the static ZLib build which requires to compile the ZLib sources a second time.

> CMake Error at Source/CMakeLists.txt:103 (add_executable):
>   Target "woof" links to target "SDL2::image" but the target was not found.

You don't have SDL2_Image installed, right?

> (I'm also personally feeling deeply uncomfortable with this automatic downloading business... :/)

It's not perfect, but at least the downloaded stuff is hash-checked. This isn't too different from a Linux distribution downloading binary packages - and since we don't have a package management system in Windows, I'd consider this good enough.

**Comment by fabiangreffrath** [comment]
@AlexMax Does `if(WIN32)` mean "a build running on Windows" or "a build targetting Windows"?

**Comment by chungy** [comment]
> You don't have SDL2_Image installed, right?

You're right, I didn't. Started working after I installed it... and just makes me wonder why we're downloading files if we require them to be installed already anyway??

Even after all that... it still fails to complete the compilation because of a missing `libopus-0.dll` file. And yes, to confirm, I am trying to build commit e80bf016c8446dcdfe4aaa0e295698a49dbdbb05

**Comment by AlexMax** [comment]
> @AlexMax Does `if(WIN32)` mean "a build running on Windows" or "a build targetting Windows"?

Targeting.  In other words, it works if you cross-compile.

> You're right, I didn't. Started working after I installed it... and just makes me wonder why we're downloading files if we require them to be installed already anyway??

You actually shouldn't need SDL_image installed.  Did you try deleting your build directory and starting fresh?  I suspect that an old cached `SDL_DIR` might have been pointing someplace it shouldn't.  The build files as they exist right now should download SDL stuff in all cases on Win32 unless you specifically override it with an `SDL_*_DIR`.

**Comment by chungy** [comment]
> You actually shouldn't need SDL_image installed. Did you try deleting your build directory and starting fresh?

I routinely do a `git clean -xfd` after or before builds. I don't leave build directories around like that.

> If it still doesn't work, paste the contents of your `CMakeCache.txt` into a pastebin and I'll take a look at it.

https://gist.github.com/chungy/b78c3ee8b16a1241b99b084141752288

**Comment by AlexMax** [comment]
> > If it still doesn't work, paste the contents of your `CMakeCache.txt` into a pastebin and I'll take a look at it.
> 
> https://gist.github.com/chungy/b78c3ee8b16a1241b99b084141752288

Judging from the cache, it looks it is finding the system libraries ahead of the ones it specifically downloads.  This should not happen unless you are passing `-DSDL_DIR` and friends to the cmake generator, as I tested it just now with my own local copy of MSYS2 with SDL libraries installed from `pacman`, and it still defaulted to the libraries it downloads itself.

Now that we download the proper libraries on Windows, you should be able to get away with a simple `cmake ..` in MSYS2.

**Comment by chungy** [comment]
>  This should not happen unless you are passing `-DSDL_DIR` and friends to the cmake generator

Neither I nor the mingw32-cmake script is passing this into it.

I guess it's suffice to say that i have no clue how to build this branch for Windows targets anymore.

**Comment by AlexMax** [comment]
Yeah, your environment seems super-messed up.  Am I to presume that you also installed `mingw64/mingw-w64-x86_64-cmake` and not `msys/cmake` and perhaps not relying on an outside CMake?

**Comment by chungy** [comment]
>  Am I to presume that you also installed `mingw64/mingw-w64-x86_64-cmake` and not `msys/cmake` and perhaps not relying on an outside CMake?

I honestly don't know what you're asking here.

**Comment by AlexMax** [comment]
> I honestly don't know what you're asking here.

There are multiple versions of CMake available in MSYS2.  You need to make sure you're installing `mingw-w64-x86_64-cmake` and not plain old `cmake`.

**Comment by chungy** [comment]
I am not using MSYS2, this is Arch Linux.

**Comment by AlexMax** [comment]
> I am not using MSYS2, this is Arch Linux.

Okay, in that case, are you using the toolchain file?  Try `-DCMAKE_TOOLCHAIN_FILE=../cmake/CrossToWin64.cmake`.

**Comment by chungy** [comment]
> Okay, in that case, are you using the toolchain file? Try `-DCMAKE_TOOLCHAIN_FILE=../cmake/CrossToWin64.cmake`.

No, I didn't even realize it's there. Works after using that... though I doubt I'll ever remember the invocation

**Comment by fabiangreffrath** [comment]
> No, I didn't even realize it's there. Works after using that... though I doubt I'll ever remember the invocation

Well, then we'll need to have it properly documented somewhere. 

@AlexMax Would you please add a short "Cross Building" paragraph to the README.md document? And would you also please not build the static ZLib anymore? Thanks!

**Comment by AlexMax** [comment]
> Well, then we'll need to have it properly documented somewhere.
> 
> @AlexMax Would you please add a short "Cross Building" paragraph to the README.md document?

Absolutely.

> And would you also please not build the static ZLib anymore? Thanks!

Not something I have direct control over - zlib's `make install` always installs the zlib and zlibstatic targets.  I might be able to come up with something else though.

**Comment by fabiangreffrath** [comment]
@AlexMax Ready for merging?

**Comment by AlexMax** [comment]
I think it is, yeah.

**Comment by fabiangreffrath** [comment]
Yeeha, thanks!

**Pull request merged**